exporter/exporterqueue/waiting_queue.go

```diff
@@ -0,0 +1,121 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package exporterqueue // import "go.opentelemetry.io/collector/exporter/exporterqueue"
+
+import (
+	"context"
+	"sync"
+
+	"go.opentelemetry.io/collector/component"
+)
+
+var waitingDonePool = sync.Pool{
+	New: func() any {
+		return &waitingDone{ch: make(chan error, 1)}
+	},
+}
+
+func newWaitingQueue[T any](sizer sizer[T], capacity int64, blocking bool, consumeFunc ConsumeFunc[T]) Queue[T] {
+	wq := &waitingQueue[T]{
+		consumeFunc: consumeFunc,
+		sizer:       sizer,
+		capacity:    capacity,
+		blocking:    blocking,
+	}
+	wq.hasMoreSpace = newCond(&wq.mu)
+	return wq
+}
+
+type waitingQueue[T any] struct {
+	component.StartFunc
+	component.ShutdownFunc
+	consumeFunc ConsumeFunc[T]
+	sizer       sizer[T]
+	capacity    int64
+
+	mu           sync.Mutex
+	hasMoreSpace *cond
```

Comment:
is this sync.Cond? I'm missing a `type cond` somewhere, or is there a missing file?

Reply:
https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/exporterqueue/cond.go

It is a fork of the sync.Cond that supports context cancellation.